### PR TITLE
feat/PRO-2446/the-exchange-app

### PR DESCRIPTION
**Pull request by RanaBug**
## Description

- Full FE design implementations of all components of the PillarX Swap App
- Implementation of getting the best offer on same chain and exchange

## How Has This Been Tested?

- This has been tested only manually

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ X ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### CODE REVIEW

File: src/apps/the-exchange-app/components/CardsSwap/CardsSwap.tsx [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+            console.error(e);
+        }
+    };
+
+    const isInitialOrder = isEqual(cardPosition, initialPosition);
+
+    const renderCards = () => (
+        <div className='flex w-full gap-4 desktop:gap-8 justify-center'>
+            <SwapReceiveCard
+                onClick={() => handleClick(isInitialOrder ? cardPosition.left : cardPosition.right)}
```
Comment (IAmKio) [inline src/apps/the-exchange-app/components/CardsSwap/CardsSwap.tsx:80]: Handle click is a little bit generic, could we make it more specific please?

File: src/apps/the-exchange-app/components/EnterAmount/EnterAmount.tsx [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+            toChainId: receiveToken?.chainId || 0,
+            toTokenDecimals: receiveToken?.decimals || 0,
+        }
+        const offer = await getBestOffer(params);
+
+        return offer;
+    }
+    
+    useEffect(() => {
+        getOffer().then((offer) => setBestOffer(offer));
```
Comment (IAmKio) [inline src/apps/the-exchange-app/components/EnterAmount/EnterAmount.tsx:39]: we'll handle catching in the next task

File: src/apps/the-exchange-app/components/EnterAmount/EnterAmount.tsx [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+
+    useEffect(() => {
+        if (type === SwapReceive.SWAP) {
+            setInputValue(amountSwap ? amountSwap.tokenAmount.toString() : '0.00');
+        } else {
+            setInputValue(amountReceive ? amountReceive.tokenAmount.toString() : '0.00');
+        }
+    }, [type, amountSwap, amountReceive]);
+
+    const handleChange = async (e: React.ChangeEvent<HTMLInputElement>) => {
```
Comment (IAmKio) [inline src/apps/the-exchange-app/components/EnterAmount/EnterAmount.tsx:51]: Handle change is too generic, could we be more descriptive here?

File: src/apps/the-exchange-app/components/ExchangeAction/ExchangeAction.tsx [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+const ExchangeAction = () => {
+    const { bestOffer, swapToken } = useContext(SwapDataContext);
+    const { addToBatch } = useGlobalTransactionsBatch();
+
+    const onClickToExchange = () => {
+        if (bestOffer) {
+            addToBatch({
+                title: 'Swap assets',
+                chainId: swapToken?.chainId || 0,
```
Comment (IAmKio) [inline src/apps/the-exchange-app/components/ExchangeAction/ExchangeAction.tsx:17]: Wondering what happens if chain is 0?!

File: src/apps/the-exchange-app/context/SwapDataProvider.tsx [hunk 5]
```diff
@@ -0,0 +1,147 @@
+import React, { createContext, useContext, useState, useEffect, useMemo } from 'react';
+import { ExchangeOffer, Token } from '@etherspot/prime-sdk/dist/sdk/data';
```
Comment (IAmKio) [inline src/apps/the-exchange-app/context/SwapDataProvider.tsx:2]: Just wondering why we're important from the dist directory directly here, I'm assuming these are types? If so they should be exported (if not then they should be, that's missing)

File: src/apps/the-exchange-app/context/SwapDataProvider.tsx [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+  const [amountReceive, setAmountReceive] = useState<AmountType | undefined>(undefined);
+  const [bestOffer, setBestOffer] = useState<ExchangeOffer | undefined>(undefined)
+
+  useEffect(() => {
+    const getAssetsSwap = async (chainId: number | undefined) => {
+      try {
+        const assets = await getPillarSwapAssets(chainId || undefined);
+        setSwapTokenData(assets);
+      } catch (error) {
+        console.error('Error fetching supported assets:', error);
```
Comment (IAmKio) [inline src/apps/the-exchange-app/context/SwapDataProvider.tsx:81]: Just a note for later, we need to handle this

File: src/apps/the-exchange-app/images/logo-unknown.png
Comment (IAmKio) [inline src/apps/the-exchange-app/images/logo-unknown.png:1]: just  note we'll replace this with Boring Avatars in future

File: src/apps/the-exchange-app/manifest.json [hunk 7]
```diff
@@ -0,0 +1,9 @@
+{
+  "title": "The Echange",
```
Comment (IAmKio) [inline src/apps/the-exchange-app/manifest.json:2]:
```suggestion
  "title": "The Exchange",
```

**Review by IAmKio — CHANGES_REQUESTED** [review]
Great work! Looking forward to the next stage of this app! Just to summarise:

- I've left the bulk of the review till the end
- Some issues are repeating and I didn't add anymore comments for them
- Could we rename the directory / ID from `the-exchange-app` to just `the-exchange` please?

Thanks and great work!

File: src/apps/the-exchange-app/components/CardsSwap/CardsSwap.tsx (hunk 1 above)
Comment (RanaBug) [reply src/apps/the-exchange-app/components/CardsSwap/CardsSwap.tsx:80]: Yes agreed, I changed it to: handleOpenTokenList

File: src/apps/the-exchange-app/components/EnterAmount/EnterAmount.tsx (hunk 2 above)
Comment (RanaBug) [reply src/apps/the-exchange-app/components/EnterAmount/EnterAmount.tsx:39]: 100% agreed, not all calls are being handled properly

File: src/apps/the-exchange-app/components/EnterAmount/EnterAmount.tsx (hunk 3 above)
Comment (RanaBug) [reply src/apps/the-exchange-app/components/EnterAmount/EnterAmount.tsx:51]: Absolutely, I changed it to: handleTokenAmount

File: src/apps/the-exchange-app/components/ExchangeAction/ExchangeAction.tsx (hunk 4 above)
Comment (RanaBug) [reply src/apps/the-exchange-app/components/ExchangeAction/ExchangeAction.tsx:17]: This needs to be better handled in the cleanup, that was a quick win

File: src/apps/the-exchange-app/context/SwapDataProvider.tsx (hunk 5 above)
Comment (RanaBug) [reply src/apps/the-exchange-app/context/SwapDataProvider.tsx:2]: Mmm... I noticed that all types from the SDK are coming from this folder

File: src/apps/the-exchange-app/context/SwapDataProvider.tsx (hunk 6 above)
Comment (RanaBug) [reply src/apps/the-exchange-app/context/SwapDataProvider.tsx:81]: 100%

File: src/apps/the-exchange-app/manifest.json (hunk 7 above)
Comment (RanaBug) [reply src/apps/the-exchange-app/manifest.json:2]: Oops

**Review by IAmKio — APPROVED** [review]
Great work 🚀

**Pull request merged**